### ISSUE TITLE
How to make a "superuser" administrator member?

### ISSUE THREAD

**Issue by PeteCoward** [dexie/Dexie.js#1755]
Using Dexie Cloud Access Control, as far as I can tell from the documentation and my small experiments, no single web authenticated user would ever be able to maintain full manage access to any potentially future stored object.

As far as I understand currently, users can be granted full manage permissions to existing realms, through roles or direct membership, and the permissions can be scoped to individual tables, but the primary beast is the realm. I can't grant an individual full manage to all objects in a given table, just a realm.

From the REST API or Cloud CLI a user could be given administrator access on all known realms at a point in time through memberships.
But any authenticated ACCESS_DB scope user (the default with the Zero config setup) could come and create a new realm and add items into any schema exposed table in those new realms.
No existing user other than the creator (and a GLOBAL_WRITE, GLOBAL_READ scope user) could see, edit or delete that data.

So to make a traditional (I'm from a Django background) superuser, I would need to implement [custom authentication](https://dexie.org/cloud/docs/db.cloud.configure()#example-integrate-custom-authentication), and request a token for that user with GLOBAL_READ and GLOBAL_WRITE.

I'm opening this issue to ask if my understanding is correct, not reporting a bug!

p.s. I'm not sure I'd want to actually have a superuser yet, I'm just trying to imagine how I would go about rebuilding an existing django / pouchdb / couchdb app in Dexie. It may well be better not to fight the paradigm. It may just be that the Zero Config access control granting ACCESS_DB to all users might not suit certain use cases.

p.p.s Would a custom authentication that granted no [scopes](https://dexie.org/cloud/docs/cli#scopes), but set up role memberships prevent any authenticated user from adding things the server does not want them to?

**Comment by dfahlander** [comment]
You're right that in the zero-authentication case, there's no way to become a superuser. You'd need to do server-to-server call to the token endpoint, using a client ID and secret to request such a token with a GLOBAL_READ and/or GLOBAL_WRITE scope. A superuser can utilize a token to query the REST API, but the sync endpoint is currently not tested to sync the entire DB as a superuser but just the realms that the user has access to as a normal or unauthorized user (unauthorized users can also sync but only from the public realm). The sync endpoint should conceptually work for a superuser also - just that the use case would probably not be to sync it to a Dexie instance (as it could easily return too much data for the client to handle in indexedDB). I might be wrong here but I would not put too much effort in testing that out right now unless I'd see a compelling use case for it.

**Comment by PeteCoward** [comment]
Thanks, the amount of data more required for a superuser over a normal user is not necessarily significant for us. In a use case where you want all users to see most of the data (it is not heavily realm-sharded like the canonical todo list example) it should not make much of a difference.

It's a good reminder though to think about that weight of data on the client, In our case over time. We did run into issue with that in our pouch/couch app eventually, more in the performance of an initial synchronisation than actual space limits on the client though.

All good, my investigations continue, but slowly!

**Comment by phil-w** [comment]
Ah, this is what I'm doing - the server-to-server thing, with the server allowing just the super-user to do whatever I need with that level of access. I've made it work, but I'm unconvinced I'm not fighting the paradigm with it to date.

**Comment by dfahlander** [comment]
> Ah, this is what I'm doing - the server-to-server thing, with the server allowing just the super-user to do whatever I need with that level of access. I've made it work, but I'm unconvinced I'm not fighting the paradigm with it to date.

The way @phil-w does it is by requesting the token using his own server-side code and there it is possible to add more scopes. This can also be achieved by configuring the fetchTokens parameter to a custom endpoint to request tokens with arbitrary claims. Currently that also means you need to implement your own authentication flow and cannot reuse the email OTP flow out of the box. But @phil-w solves that by leaving the built in authentication but then offers another api that elevates the token.

I wouldn't say that it's fighting the paradigm but rather find ways around the current limitations.

**Comment by phil-w** [comment]
This is true. 

I also used auth.js as an auth provider, and _that worked with Dexie Cloud very well_ and was an easy integrate. 

I decided that way wasn't useful for me, because auth.js is OpenAuth2 based (a good thing), OpenAuth doesn't work well for API authentication when you have mobile devices as well as PWAs to deal with. It's annoying as it's fine if it's just a browser, but mobile is a killer with "in app browsers" being necessary as it's essentially a HTML thing. That one is definitely working against a paradigm!  

Hence I'm working with Dexie Cloud's own auth, which avoids that hassle.
